### PR TITLE
Fix Diagnostic layer packaging and wiring + Added #596

### PR DESCRIPTION
**Pull request by mathieucarbou**

### CODE REVIEW

File: diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateMBeanImpl.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  public LogicalServerStateMBeanImpl() throws NotCompliantMBeanException {
+    this(new JMXSubsystem());
+  }
+
+  public LogicalServerStateMBeanImpl(JMXSubsystem jmxSubsystem) throws NotCompliantMBeanException {
+    super(LogicalServerStateMbean.class, false);
+    this.subsystem = jmxSubsystem;
+  }
+
+  public void init() {
```
Comment (saurabhagas) [inline diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateMBeanImpl.java]: We don't need this `init` method. The contents can be moved to the constructor.

File: diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateServiceProvider.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.terracotta.entity.ServiceProvider;
+import org.terracotta.entity.ServiceProviderConfiguration;
+
+import javax.management.NotCompliantMBeanException;
+import java.util.Collection;
+import java.util.Collections;
+
+@BuiltinService
+public class LogicalServerStateServiceProvider implements ServiceProvider {
+  private volatile LogicalServerStateMBeanImpl detailedServerState;
```
Comment (saurabhagas) [inline diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateServiceProvider.java]: variable name needs to be updated.

File: diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateServiceProvider.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Override
+  public <T> T getService(long l, ServiceConfiguration<T> serviceConfiguration) {
+    Class<T> serviceType = serviceConfiguration.getServiceType();
+    return serviceType != org.terracotta.diagnostic.server.api.extension.LogicalServerStateProvider.class ? null : serviceType.cast(detailedServerState);
+  }
+
+  @Override
+  public Collection<Class<?>> getProvidedServiceTypes() {
+    return Collections.singletonList(LogicalServerStateMBeanImpl.class);
```
Comment (saurabhagas) [inline diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateServiceProvider.java]: Needs to be replaced by interface

File: diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateServiceProvider.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      return true;
+    } catch (NotCompliantMBeanException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public <T> T getService(long l, ServiceConfiguration<T> serviceConfiguration) {
+    Class<T> serviceType = serviceConfiguration.getServiceType();
+    return serviceType != org.terracotta.diagnostic.server.api.extension.LogicalServerStateProvider.class ? null : serviceType.cast(detailedServerState);
```
Comment (saurabhagas) [inline diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateServiceProvider.java]: We can replace the full import with a qualified import.

File: diagnostic/service/src/test/java/org/terracotta/diagnostic/server/extensions/DetailedServerStateTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import static org.mockito.Mockito.when;
+import static org.terracotta.diagnostic.common.DiagnosticConstants.MBEAN_CONSISTENCY_MANAGER;
+import static org.terracotta.diagnostic.common.DiagnosticConstants.MBEAN_SERVER;
+import static org.terracotta.diagnostic.common.DiagnosticConstants.MESSAGE_INVALID_JMX;
+import static org.terracotta.diagnostic.model.LogicalServerState.ACTIVE;
+import static org.terracotta.diagnostic.model.LogicalServerState.ACTIVE_RECONNECTING;
+import static org.terracotta.diagnostic.model.LogicalServerState.ACTIVE_SUSPENDED;
+import static org.terracotta.diagnostic.model.LogicalServerState.UNKNOWN;
+
+public class DetailedServerStateTest {
```
Comment (saurabhagas) [inline diagnostic/service/src/test/java/org/terracotta/diagnostic/server/extensions/DetailedServerStateTest.java]: Name needs to be updated

File: diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateMBeanImpl.java (hunk 1 above)
Comment (mathieucarbou) [reply diagnostic/service/src/main/java/org/terracotta/diagnostic/server/extensions/LogicalServerStateMBeanImpl.java]: we need to separate object construction from JMX registration.

**Comment by mathieucarbou** [comment]
@saurabhagas : test do not pass:
```
org.terracotta.diagnostic.client.DiagnosticOperationExecutionException: Invalid JMX call:hasServiceInterface org.terracotta:name=DiagnosticRequestHandler

	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.execute(DiagnosticServiceImpl.java:211)
	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.execute(DiagnosticServiceImpl.java:180)
	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.invokeWithArg(DiagnosticServiceImpl.java:136)
	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.getProxy(DiagnosticServiceImpl.java:82)
	at org.terracotta.dynamic_config.system_tests.diagnostic.TopologyServiceIT.test_getPendingTopology(TopologyServiceIT.java:62)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```
Looking... The diagnostic layer is not made available...

**Comment by saurabhagas** [comment]
> @saurabhagas : test do not pass:
> 
> ```
> org.terracotta.diagnostic.client.DiagnosticOperationExecutionException: Invalid JMX call:hasServiceInterface org.terracotta:name=DiagnosticRequestHandler
> 
> 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.execute(DiagnosticServiceImpl.java:211)
> 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.execute(DiagnosticServiceImpl.java:180)
> 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.invokeWithArg(DiagnosticServiceImpl.java:136)
> 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.getProxy(DiagnosticServiceImpl.java:82)
> 	at org.terracotta.dynamic_config.system_tests.diagnostic.TopologyServiceIT.test_getPendingTopology(TopologyServiceIT.java:62)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
> 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
> 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
> 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> 
> Looking... The diagnostic layer is not made available...

Is the word "Mbean" case sensitive? I see mostly "MBean" names..

**Comment by mathieucarbou** [comment]
> > @saurabhagas : test do not pass:
> > ```
> > org.terracotta.diagnostic.client.DiagnosticOperationExecutionException: Invalid JMX call:hasServiceInterface org.terracotta:name=DiagnosticRequestHandler
> > 
> > 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.execute(DiagnosticServiceImpl.java:211)
> > 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.execute(DiagnosticServiceImpl.java:180)
> > 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.invokeWithArg(DiagnosticServiceImpl.java:136)
> > 	at org.terracotta.diagnostic.client.DiagnosticServiceImpl.getProxy(DiagnosticServiceImpl.java:82)
> > 	at org.terracotta.dynamic_config.system_tests.diagnostic.TopologyServiceIT.test_getPendingTopology(TopologyServiceIT.java:62)
> > 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> > 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> > 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> > 	at java.lang.reflect.Method.invoke(Method.java:498)
> > 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> > 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> > 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> > 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> > 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> > 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
> > 	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
> > 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
> > 	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
> > 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> > 	at java.lang.Thread.run(Thread.java:748)
> > ```
> > 
> > 
> > Looking... The diagnostic layer is not made available...
> 
> Is the word "Mbean" case sensitive? I see mostly "MBean" names..

there is another problem too. packaging problem...

**Comment by mathieucarbou** [comment]
@saurabhagas : packaging issues are fixed. 
- I got other problem with maven not correctly picking up the right manifest so there were classpath issues.
- I was able to get rid of the MBean interface and use the one from the provider. The issue we had was because of the kit packaging - there were a missing lib on server
- All Galvan MnM tests passed
- Angela dynamic config tests are running and all green at the moment.

**Comment by mathieucarbou** [comment]
@saurabhagas : I passed the tests several times locally, but on Azure, I do not understand why, I get: 
```
Fatal Terracotta startup exception:

 Address already in use.  Please make sure the server isn't already running or choose a different port.
```
The PR is good and also unrelated to that since the failure is in `AutoActivateNewPassive1x2IT`.

=> So we can merge.

**Pull request merged**